apis/projectcontour/v1alpha1/contourconfig.go

```diff
@@ -22,6 +25,396 @@ import (
+
+	// FallbackCertificate defines the namespace/name of the Kubernetes secret to
+	// use as fallback when a non-SNI request is received.
+	FallbackCertificate NamespacedName `json:"fallbackCertificate,omitempty"`
+}
+
+// LeaderElectionParameters holds the config bits for leader election
+// inside the  config file.
+type LeaderElectionParameters struct {
+	LeaseDuration         string         `json:"leaseDuration,omitempty"`
+	RenewDeadline         string         `json:"renewDeadline,omitempty"`
+	RetryPeriod           string         `json:"retryPeriod,omitempty"`
+	Configmap             NamespacedName `json:"configmap,omitempty"`
+	DisableLeaderElection bool           `json:"disableLeaderElection,omitempty"`
+}
+
+// NetworkParameters hold various configurable network values.
+type NetworkParameters struct {
+	// XffNumTrustedHops defines the number of additional ingress proxy hops from the
+	// right side of the x-forwarded-for HTTP header to trust when determining the origin
+	// client’s IP address.
+	//
+	// See https://www.envoyproxy.io/docs/envoy/v1.17.0/api-v3/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto?highlight=xff_num_trusted_hops
+	// for more information.
+	XffNumTrustedHops uint32 `json:"numTrustedHops,omitempty"`
+
+	// Configure the port used to access the Envoy Admin interface.
+	// If configured to port "0" then the admin interface is disabled.
+	EnvoyAdminPort int `json:"adminPort,omitempty"`
+}
+
+// RateLimitService defines properties of a global Rate Limit Service.
+type RateLimitService struct {
+	// ExtensionService identifies the extension service defining the RLS,
+	// formatted as <namespace>/<name>.
+	ExtensionService string `json:"extensionService,omitempty"`
+
+	// Domain is passed to the Rate Limit Service.
+	Domain string `json:"domain,omitempty"`
+
+	// FailOpen defines whether to allow requests to proceed when the
+	// Rate Limit Service fails to respond with a valid rate limit
+	// decision within the timeout defined on the extension service.
+	FailOpen bool `json:"failOpen,omitempty"`
+
+	// EnableXRateLimitHeaders defines whether to include the X-RateLimit
+	// headers X-RateLimit-Limit, X-RateLimit-Remaining, and X-RateLimit-Reset
+	// (as defined by the IETF Internet-Draft linked below), on responses
+	// to clients when the Rate Limit Service is consulted for a request.
+	//
+	// ref. https://tools.ietf.org/id/draft-polli-ratelimit-headers-03.html
+	EnableXRateLimitHeaders bool `json:"enableXRateLimitHeaders,omitempty"`
+}
+
+// PolicyParameters holds default policy used if not explicitly set by the user
+type PolicyParameters struct {
+	// RequestHeadersPolicy defines the request headers set/removed on all routes
+	RequestHeadersPolicy HeadersPolicy `json:"requestHeaders,omitempty"`
+
+	// ResponseHeadersPolicy defines the response headers set/removed on all routes
+	ResponseHeadersPolicy HeadersPolicy `json:"responseHeaders,omitempty"`
+}
```

Comment:
I'm thinking these fields could move into `HTTPProxyConfig` since (a) `Policy` is a pretty vague name and (b) these only get applied to `HTTPProxy` config anyway. I guess though if we ever wanted to apply these to Ingress/Gateway routes, that wouldn't make sense. Thoughts?

Reply:
Heh OK just saw https://github.com/projectcontour/contour/pull/4018 so probably doesn't make sense to move this into `HTTPProxy`. If we can think of a more descriptive name than `Policy`, great, otherwise, not a big deal.